### PR TITLE
Addition of weighted selection to custom providers

### PR DESCRIPTION
**Pull request by SamuelNeceda**
As for the [Issue](https://github.com/datafaker-net/datafaker/issues/1314), I have implemented the capability for weighted selection with custom providers.

datafaker-net/datafaker#1314

### CODE REVIEW

**Review by kingthorin — COMMENTED** [review]
What happens in the case of an empty or non-double weight?

(I haven't pulled the code and tried).

I'd also suggest creating a test YAML file and testing as you see users actually using it. (Which can likely be re-used for documentation later)

**Comment by SamuelNeceda** [comment]
@kingthorin , great catch on the edge cases for weights! I’ve implemented a solution that should address the necessary scenarios.

Additionally, in [this file](https://github.com/datafaker-net/datafaker/pull/1414/files#diff-6b867d96f22fbd6485e9d317a9dffd8a7870afd9c37656cdf53b19ea5a5f8d58), I demonstrated the weighted selection with a custom hardcoded provider.

Currently, I’m working on implementing weighted selection using a custom provider that loads data from a YAML file. I’ve created the file, but I’m facing challenges loading the data to apply the weighted selection. Following the logic from the ant() method, it looks like the key changes might be needed in AbstractProvider and FakeValueService classes, but I'm unsure how to proceed. Could you please provide any recommendations?

**Review by kingthorin — DISMISSED** [review]

File: src/main/java/net/datafaker/service/RandomService.java [hunk 1]
```diff
@@ -141,6 +145,84 @@ public Random getRandomInternal() {
...
+        if (weight <= 0 || Double.isNaN(weight) || Double.isInfinite(weight)) {
+            throw new IllegalArgumentException("Weight must be a positive number and cannot be NaN or infinite");
+        }
+    }
+
+    private double calculateTotalWeight(List<Map<String, Object>> items) {
+        return items.stream()
+            .mapToDouble(item -> (Double) item.get(WEIGHT_KEY))
+            .sum();
+    }
```
Comment (snuyanzin) [inline src/main/java/net/datafaker/service/RandomService.java:210]: How will it work if 
1. there are N weights each of them is Double.MAX_VALUE - 1 or something like that, are we protected against overflow?
2. what if there are custom fields with same name however with different meaning in same provider and even number type? Will there at least any warning appear  or how should user be warned about this (without hours of debugging)

File: src/main/java/net/datafaker/service/RandomService.java [hunk 2]
```diff
@@ -141,6 +145,84 @@ public Random getRandomInternal() {
...
+
+    private <T> T selectWeightedElement(List<Map<String, Object>> items, double randomValue) {
+        double currentWeight = 0.0;
+
+        for (Map<String, Object> item : items) {
+            currentWeight += (Double) item.get(WEIGHT_KEY);
+            if (randomValue < currentWeight) {
+                return (T) item.get(VALUE_KEY);
+            }
+        }
```
Comment (snuyanzin) [inline src/main/java/net/datafaker/service/RandomService.java:220]: is there a reason to do this on each iteration?
IIUC data in file is persistent then we could sort them once including all unboxing things (not sure if need them...) only once and have something like array of doubles
then just use something like adopted binary search

or did I miss anything here?

File: src/test/java/net/datafaker/utils/TestConstants.java
Comment (kingthorin) [inline src/test/java/net/datafaker/utils/TestConstants.java:1]: Is there a reason these can't be public in the appropriate test class?

Comment (SamuelNeceda) [reply src/test/java/net/datafaker/utils/TestConstants.java:1]: I moved the constants to the appropriate test class and removed TestConstants.java. Initially, they were placed in a separate utils package to make them accessible across multiple classes, but since they are only used here, keeping them in the test class simplifies the structure.

File: src/main/java/net/datafaker/service/RandomService.java (hunk 1 above)
Comment (SamuelNeceda) [reply src/main/java/net/datafaker/service/RandomService.java]: 1. I appreciate your insight regarding potential overflow. While weights are typically not expected to be extremely large, I've implemented an iterative approach to ensure that adding each weight is checked against Double.MAX_VALUE. This helps prevent overflow in cases where unusually large weights might occur.
2. I believe that values should be unique when performing weighted selection. Therefore, while weights can be the same, duplicates in the value parameter trigger an exception that informs the user of the specific duplicate value. This approach seems appropriate to ensure the integrity of the selection process.

File: src/main/java/net/datafaker/service/RandomService.java (hunk 2 above)
Comment (SamuelNeceda) [reply src/main/java/net/datafaker/service/RandomService.java]: Thank you for the suggestion. I implemented the proposed optimization of sorting and preprocessing the cumulative weights once, which improved the efficiency.

File: src/main/java/net/datafaker/service/RandomService.java [hunk 3]
```diff
@@ -1,14 +1,24 @@
...
+    private static final String WEIGHT_KEY = "weight";
+    private static final String VALUE_KEY = "value";
     private static final char[] HEX_UP = "0123456789ABCDEF".toCharArray();
     private static final char[] HEX_LOWER = "0123456789abcdef".toCharArray();
     private static final Random SHARED_RANDOM = new Random();
     private final Random random;
 
+    private double[] cumulativeWeights;
+    private Object[] values;
+
```
Comment (snuyanzin) [inline src/main/java/net/datafaker/service/RandomService.java:21]: Definitively it should be outside of this class
Otherwise it will slow down all other not weight related cases

There should be a weight provider dedicated interface/class instead this very generic

File: src/main/java/net/datafaker/service/RandomService.java [hunk 4]
```diff
@@ -141,6 +151,101 @@ public Random getRandomInternal() {
...
+        if (!values.add(value)) {
+            throw new IllegalArgumentException("Duplicate value found: " + value + ". Values must be unique.");
+        }
+    }
+
+    private void validateItem(Map<String, Object> item) {
+        if (item == null || item.isEmpty()) {
+            throw new IllegalArgumentException("Item cannot be null or empty");
+        }
+        if (!item.containsKey(WEIGHT_KEY) || !item.containsKey(VALUE_KEY)) {
```
Comment (snuyanzin) [inline src/main/java/net/datafaker/service/RandomService.java:199]: same here about equals/hashcode

File: src/main/java/net/datafaker/service/RandomService.java (hunk 1 above)
Comment (snuyanzin) [reply src/main/java/net/datafaker/service/RandomService.java]: > While weights are typically not expected to be extremely large,

if we are talking about project in open source which is used then any input should be expected if it passes all existing validation

Comment (SamuelNeceda) [reply src/main/java/net/datafaker/service/RandomService.java]: Sure, I got your point. I provided the guard for overflow.

File: src/main/java/net/datafaker/service/RandomService.java [hunk 5]
```diff
@@ -141,6 +151,101 @@ public Random getRandomInternal() {
...
+
+        for (var item : items) {
+            validateItem(item);
+            assertUniqueValues(item, uniqueValues);
+        }
+    }
+
+    private void assertUniqueValues(Map<String, Object> item, Set<Object> values) {
+        Object value = item.get(VALUE_KEY);
+        if (!values.add(value)) {
```
[the remark this answers is not in the document]
Comment (SamuelNeceda) [reply src/main/java/net/datafaker/service/RandomService.java:190]: Do you have some suggestions on how to refactor it?

File: src/main/java/net/datafaker/service/RandomService.java (hunk 4 above)
Comment (SamuelNeceda) [reply src/main/java/net/datafaker/service/RandomService.java:199]: Sorry, but I might not get this. I am checking if each item contains weight and value keys defined as strings. If for some reason they do not, I am directly throwing an exception.

File: src/main/java/net/datafaker/service/RandomService.java (hunk 3 above)
Comment (SamuelNeceda) [reply src/main/java/net/datafaker/service/RandomService.java]: I've moved all the logic for the weight selector to the WeightedRandomSelector class. Currently, I’ve placed it in the service package, as it seems to fit best there, though I believe you initially suggested putting it in the providers package?

Additionally, based on the issue description, WeightedRandomSelector appears intended mainly for custom providers. At this point, no provider has data in the format required for this functionality. Should I register it only to be used with custom providers, or do you have a different solution in mind?

**Comment by SamuelNeceda** [comment]
@kingthorin, could you please review the PR? I'd like to wrap it up or address any potential comments. Thanks!

**Review by kingthorin — APPROVED** [review]

File: src/main/java/net/datafaker/service/WeightedRandomSelector.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+import java.util.Set;
+
+/**
+ * A utility class for selecting a random element from a list based on assigned weights.
+ **/
+public class WeightedRandomSelector {
+    private static final String WEIGHT_KEY = "weight";
+    private static final String VALUE_KEY = "value";
+    private double[] cumulativeWeights;
+    private Object[] values;
```
Comment (snuyanzin) [inline src/main/java/net/datafaker/service/WeightedRandomSelector.java:17]: How about several (at least 2 threads) working with this
i guess they will concurrently change same class fields and inpact each other

**Review by snuyanzin — CHANGES_REQUESTED** [review]
concurrency issue shouldn't be introduced in base classes

File: src/main/java/net/datafaker/service/WeightedRandomSelector.java (hunk 6 above)
Comment (SamuelNeceda) [reply src/main/java/net/datafaker/service/WeightedRandomSelector.java]: @snuyanzin , with my last changes, I made the class stateless. Please, review.

File: src/main/java/net/datafaker/providers/base/ProviderRegistration.java [hunk 7]
```diff
@@ -150,6 +151,8 @@ <PR extends ProviderRegistration, AP extends AbstractProvider<PR>> AP getProvide
 
     RandomService random();
 
+    WeightedRandomSelector weightedRandomSelector();
```
Comment (snuyanzin) [inline src/main/java/net/datafaker/providers/base/ProviderRegistration.java:154]: what is the reason to have it here?
I would rather keep it completely separate

File: src/main/java/net/datafaker/service/WeightedRandomSelector.java (hunk 6 above)
Comment (snuyanzin) [reply src/main/java/net/datafaker/service/WeightedRandomSelector.java]: i see, however it looks like now again refilling of arrays is happening on each iteration or did I miss anything?

File: src/main/java/net/datafaker/service/WeightedRandomSelector.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+     * @return A randomly selected element based on its weight.
+     * @throws IllegalArgumentException if:
+     *                                  - the list is null or empty,
+     *                                  - any item in the list is null or empty,
+     *                                  - the item does not contain 'weight' or 'value' keys,
+     *                                  - any weight is null, non-positive, NaN or infinite,
+     *                                  - any values in the list are not unique or null,
+     *                                  - the sum of weights exceeds Double.MAX_VALUE.
+     */
+    public <T> T select(List<Map<String, Object>> items) {
```
Comment (snuyanzin) [inline src/main/java/net/datafaker/service/WeightedRandomSelector.java:43]: looks like most of the methods here do not work with object state => could be turned to `static`

File: src/test/weightedAnts.yml [hunk 9]
```diff
@@ -0,0 +1,10 @@
+en:
+  faker:
+    insectsfromfile:
+      weightedAnts:
+        - value: "Driver ant"
+          weight: 6.0
+        - value: "Fire ant"
+          weight: 3.0
+        - value: "Harvester ant"
+          weight: 1.0
```
Comment (snuyanzin) [inline src/test/weightedAnts.yml:10]: With simple example it works
however what it we have more complex structure like 
https://github.com/datafaker-net/datafaker/blob/73ae1267e053dd9cb93039b1c721fdf0d282996b/src/main/resources/en/planet.yml#L14-L71

File: src/main/java/net/datafaker/service/WeightedRandomSelector.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    public WeightedRandomSelector(Random random) {
+        this.random = random != null ? random : new Random();
+    }
+
+    public Random getRandom() {
+        return random;
+    }
+
+    /**
+     * Returns a weighted random element from the given list, where each element is represented as a Map
```
Comment (snuyanzin) [inline src/main/java/net/datafaker/service/WeightedRandomSelector.java:27]: Am I right that the only way to change weight is change file and rebuild everything?

File: src/main/java/net/datafaker/service/WeightedRandomSelector.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    private void validateWeight(Object weightObj) {
+        if (!(weightObj instanceof Double weight)) {
+            throw new IllegalArgumentException("Weight must be a non-null Double");
+        }
+        if (weight <= 0 || Double.isNaN(weight) || Double.isInfinite(weight)) {
+            throw new IllegalArgumentException("Weight must be a positive number and cannot be NaN or infinite");
+        }
+    }
+
+    public double[] preprocessItems(List<Map<String, Object>> items, Object[] values) {
```
Comment (snuyanzin) [inline src/main/java/net/datafaker/service/WeightedRandomSelector.java:99]: I wonder why it is public?

File: src/main/java/net/datafaker/service/WeightedRandomSelector.java [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+        return cumulativeWeights;
+    }
+
+    public <T> T selectWeightedElement(double randomValue, double[] cumulativeWeights, Object[] values) {
+        int index = Arrays.binarySearch(cumulativeWeights, randomValue);
+        index = (index < 0) ? -index - 1 : index;
+
+        if (index >= cumulativeWeights.length) {
+            index = cumulativeWeights.length - 1;
+        }
```
Comment (snuyanzin) [inline src/main/java/net/datafaker/service/WeightedRandomSelector.java:122]:
```suggestion
        index = Math.max(index, cumulativeWeights.length - 1);
```
i think it would be shorter

File: src/main/java/net/datafaker/service/WeightedRandomSelector.java [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+        if (!values.add(value)) {
+            throw new IllegalArgumentException("Duplicate value found: " + value + ". Values must be unique.");
+        }
+    }
+
+    private void validateItem(Map<String, Object> item) {
+        if (item == null || item.isEmpty()) {
+            throw new IllegalArgumentException("Item cannot be null or empty");
+        }
+        if (!item.containsKey(WEIGHT_KEY) || !item.containsKey(VALUE_KEY)) {
```
Comment (snuyanzin) [inline src/main/java/net/datafaker/service/WeightedRandomSelector.java:77]: are we checking separately existing of keys and non null values only for the sake of different error messages>

**Comment by SamuelNeceda** [comment]
Thank you for the feedback @snuyanzin
At this point, I’m not in a position to fully address the comments. Initially, I expected this issue to involve only minor changes, but it has become more complex than anticipated, and I’m unable to foresee all edge cases. Would it be possible for someone else to take it over, or should we consider creating a separate issue to track this?

**Comment by snuyanzin** [comment]
with current architecture this topic is very complex that's why there are so many questions.
My proposal is to not touch base classes at all and have something as a proof of concept (POC) which could be accessible only on demand and not used by default

**Comment by SamuelNeceda** [comment]
So the recommendation is to keep this as a POC only. What’s the next step from here? I’d like to ensure the work I've done isn't lost, even though I'm unable to complete the entire issue as intended. Any suggestions @kingthorin @snuyanzin?

**Comment by kingthorin** [comment]
I don't mind "finishing" it. However I believe all that might be left to bring it in for POC purposes is to disconnect it from the faker or value service. Right @snuyanzin? 

Just let WeightedRandomSelector be instantiated and do all the work for this functionality while people test it and improve it.

**Comment by snuyanzin** [comment]
yes, kind of
while all major issues are not solve we shouldn't introduce any change in classes like `BaseFaker.java`, `ProviderRegistration.java`, `FakeValuesService.java`

File: src/main/java/net/datafaker/service/WeightedRandomSelector.java (hunk 12 above)
Comment (SamuelNeceda) [reply src/main/java/net/datafaker/service/WeightedRandomSelector.java:122]: Proposed change will always set index to `cumulativeWeights.length - 1` if index is less than `cumulativeWeights.length - 1`, which is not the intended behavior. The intent is to clamp the value only when index is greater than or equal to `cumulativeWeights.length`

File: src/main/java/net/datafaker/providers/base/ProviderRegistration.java (hunk 7 above)
Comment (SamuelNeceda) [reply src/main/java/net/datafaker/providers/base/ProviderRegistration.java]: as proposed, it was removed from `ProviderRegistration.java`

File: src/main/java/net/datafaker/service/WeightedRandomSelector.java (hunk 8 above)
Comment (SamuelNeceda) [reply src/main/java/net/datafaker/service/WeightedRandomSelector.java:39]: Relevant methods were turned to static and private

File: src/test/weightedAnts.yml (hunk 9 above)
Comment (SamuelNeceda) [reply src/test/weightedAnts.yml:10]: Since it was agreed that no changes should be introduced to `FakeValuesService.java`, I removed the `fetchWeightedList` method. Consequently, interacting with YAML files is no longer supported so I have also removed the example YAML file.

File: src/main/java/net/datafaker/service/WeightedRandomSelector.java (hunk 10 above)
Comment (SamuelNeceda) [reply src/main/java/net/datafaker/service/WeightedRandomSelector.java:22]: The weights are provided as part of the input items list and are processed at runtime. Changing weights does not require rebuilding the application but simply involves modifying the input list or the file (if applicable) and rerunning the program.

File: src/main/java/net/datafaker/service/WeightedRandomSelector.java (hunk 11 above)
Comment (SamuelNeceda) [reply src/main/java/net/datafaker/service/WeightedRandomSelector.java]: changed

File: src/main/java/net/datafaker/service/WeightedRandomSelector.java (hunk 13 above)
Comment (SamuelNeceda) [reply src/main/java/net/datafaker/service/WeightedRandomSelector.java:79]: I split the condition into separate checks with distinct error messages.

**Comment by SamuelNeceda** [comment]
@kingthorin @snuyanzin 
I have removed all changes from the mentioned classes, ensuring that the WeightedRandomSelector can now be instantiated independently.

Additionally, I have updated the documentation and examples to emphasize that weighted random selection is still in the Proof of Concept (POC) stage.

I believe these changes address your concerns, and the PR is now ready for merging. Any future suggestions for improvements will be considered and addressed in subsequent Issues, ...

File: docs/documentation/custom-providers.md [hunk 14]
```diff
@@ -31,10 +31,21 @@ Create a custom provider of data:
...
         }
-    }
-    ```
 
+        public String weightedInsectName() {
+            List<Map<String, Object>> insects = List.of(
+                Map.of("value", "Driver ant", "weight", 6.0),
+                Map.of("value", "Fire ant", "weight", 3.0),
+                Map.of("value", "Harvester ant", "weight", 1.0)
+            );
```
Comment (snuyanzin) [inline docs/documentation/custom-providers.md:40]: It would be better to follow same way as existing example and place all the example data in constant like above rather than generating it for each method invocation

File: docs/documentation/custom-providers.md [hunk 15]
```diff
@@ -31,10 +31,21 @@ Create a custom provider of data:
...
+
+            WeightedRandomSelector selector = new WeightedRandomSelector(new Random());
```
Comment (snuyanzin) [inline docs/documentation/custom-providers.md:42]: Also I guess we don't need to create a new selector per getter invocation and can reuse it

File: src/main/java/net/datafaker/service/WeightedRandomSelector.java (hunk 8 above)
Comment (snuyanzin) [reply src/main/java/net/datafaker/service/WeightedRandomSelector.java:39]: how about this method
```java
<T> T select(List<Map<String, Object>> items) {
```

why can not it be switched to `static`?

File: src/main/java/net/datafaker/service/WeightedRandomSelector.java [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        if (valueObj == null) {
+            throw new IllegalArgumentException("Value cannot be null");
+        }
+    }
+
+    private static void validateWeight(Object weightObj) {
+        if (!(weightObj instanceof Double weight)) {
+            throw new IllegalArgumentException("Weight must be a non-null Double");
+        }
+        if (weight <= 0 || Double.isNaN(weight) || Double.isInfinite(weight)) {
```
Comment (snuyanzin) [inline src/main/java/net/datafaker/service/WeightedRandomSelector.java:96]: out of curiosity: why do we consider `weight = 0` as an error?
IMHO it might be a valid way to turn off this value while the total must be positive

File: src/test/java/net/datafaker/service/WeightedRandomSelectorTest.java [hunk 17]
```diff
@@ -0,0 +1,326 @@
...
+    void testWeightedArrayElement_withInfiniteWeight(WeightedRandomSelector weightedRandomSelector) {
+        List<Map<String, Object>> items = List.of(
+            Map.of(VALUE_KEY, ELEMENT_1, WEIGHT_KEY, Double.POSITIVE_INFINITY),
+            Map.of(VALUE_KEY, ELEMENT_2, WEIGHT_KEY, ELEMENT_2_WEIGHT)
+        );
+
+        assertThatThrownBy(() -> weightedRandomSelector.select(items))
+            .isInstanceOf(IllegalArgumentException.class)
+            .hasMessage("Weight must be a positive number and cannot be NaN or infinite");
+    }
```
Comment (snuyanzin) [inline src/test/java/net/datafaker/service/WeightedRandomSelectorTest.java:168]: It looks like all these tests (my be some more) could be replaced with one parameterized test
since we have input of key/value sequence, exception and error message

File: docs/documentation/custom-providers.md (hunk 15 above)
Comment (SamuelNeceda) [reply docs/documentation/custom-providers.md]: done

File: src/main/java/net/datafaker/service/WeightedRandomSelector.java (hunk 16 above)
Comment (SamuelNeceda) [reply src/main/java/net/datafaker/service/WeightedRandomSelector.java]: My thought was that a weight of 0 implies that the corresponding item cannot be selected. Since WeightedRandomSelector is designed to assign probabilities proportional to weights, including a weight of 0 would lead to unnecessary processing for an element that cannot be selected. But I refactored the code to enable also 0 weight, allowing weights of 0 to be a legitimate way to "turn off" certain elements while still keeping them in the configuration

File: src/test/java/net/datafaker/service/WeightedRandomSelectorTest.java (hunk 17 above)
Comment (SamuelNeceda) [reply src/test/java/net/datafaker/service/WeightedRandomSelectorTest.java]: Good point, I refactored a whole test class to use a parameterized tests.

**Comment by SamuelNeceda** [comment]
Thanks for the comments @snuyanzin , I have already addressed them.

File: src/test/java/net/datafaker/service/WeightedRandomSelectorTest.java [hunk 18]
```diff
@@ -0,0 +1,273 @@
...
+    public static final double ELEMENT_2_WEIGHT = 2.0;
+    public static final double ELEMENT_3_WEIGHT = 3.0;
+    public static final double NEGATIVE_WEIGHT = -1.0;
+
+    /**
+     * Tests for exception scenarios using parameterized tests.
+     */
+    @ParameterizedTest
+    @MethodSource("exceptionCasesProvider")
+    void testWeightedArrayElement_exceptions(List<Map<String, Object>> items, String expectedMessage) {
```
Comment (snuyanzin) [inline src/test/java/net/datafaker/service/WeightedRandomSelectorTest.java:34]: Can you rearrange a bit methods within this class
in a way: methods with different test annotations first
all other methods after that

Comment (snuyanzin) [inline src/test/java/net/datafaker/service/WeightedRandomSelectorTest.java:34]:
```suggestion
    void testWeightedArrayElementExceptions(List<Map<String, Object>> items, String expectedMessage) {
```
lets's do not mix different notations

Comment (SamuelNeceda) [reply src/test/java/net/datafaker/service/WeightedRandomSelectorTest.java]: refactored

File: src/main/java/net/datafaker/service/WeightedRandomSelector.java [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+     * @return A randomly selected element based on its weight.
+     * @throws IllegalArgumentException if:
+     *                                  - the list is null or empty,
+     *                                  - any item in the list is null or empty,
+     *                                  - the item does not contain 'weight' or 'value' keys,
+     *                                  - any weight is null, non-positive, NaN or infinite,
+     *                                  - any values in the list are not unique or null,
+     *                                  - the sum of weights exceeds Double.MAX_VALUE.
+     */
+    public <T> T select(List<Map<String, Object>> items) {
```
Comment (snuyanzin) [inline src/main/java/net/datafaker/service/WeightedRandomSelector.java:39]:
```suggestion
    public static <T> T select(List<Map<String, Object>> items) {
```
this seems not addressed yet
not sure if we need it public

**Review by snuyanzin — APPROVED** [review]

**Review by kingthorin — APPROVED** [review]
